src/main/java/org/hydev/mcpm/client/database/fetcher/LocalDatabaseFetcher.java

```diff
@@ -0,0 +1,192 @@
+import org.apache.hc.client5.http.fluent.Request;
+import org.apache.hc.core5.http.ClassicHttpResponse;
+import org.apache.hc.core5.http.HttpEntity;
+import org.hydev.mcpm.client.models.Database;
+import org.hydev.mcpm.utils.HashUtils;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.NoSuchAlgorithmException;
+
+/**
+ * A database fetcher that has a local persistent cache along and a network fallback.
+ */
+public class LocalDatabaseFetcher implements DatabaseFetcher {
+    private final URI host;
+    private final Path cacheDirectory;
+
+    private Database localDatabase;
+
+    public static final String HASH_FILE_NAME = "db.hash";
+    public static final String DATABASE_FILE_NAME = "db";
+
+    public static final String USER_AGENT = "MCPM Client";
+
+    /**
+     * Creates a database fetcher that syncs with a mirror.
+     * The cache directory will default to .mcpm
+     *
+     * @param host The host URL MCPM server mirror that will be used to request database files.
+     *             Example: "http://mcpm.hydev.com"
+     */
+    public LocalDatabaseFetcher(URI host) {
+        this(host, Path.of(".mcpm/"));
+    }
+
+    /**
+     * Creates a database fetcher that syncs with a mirror.
+     *
+     * @param host The host URL MCPM server mirror that will be used to request database files.
+     *             Example: "http://mcpm.hydev.com"
+     * @param cacheDirectory The directory that the cached database and hash file will be stored (as child files).
+     */
+    public LocalDatabaseFetcher(URI host, Path cacheDirectory) {
+        this.host = host;
+        this.cacheDirectory = cacheDirectory;
+    }
+
+    private Request requestTo(String path) {
+        return Request
+            .get(URI.create(host.toString() + "/" + path))
+            .addHeader("Host", host.getHost())
+            .addHeader("User-Agent", USER_AGENT)
+            .addHeader("Accepts", "application/json");
+    }
+
+    @Nullable
+    private String fetchHostHash() {
+        try {
+            var response = requestTo(HASH_FILE_NAME).execute();
+
+            return response.returnContent().asString().trim();
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    @Nullable
+    private String fetchLocalHash() {
+        try {
+            return Files.readString(cacheDirectory.resolve(DATABASE_FILE_NAME));
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    private boolean localDatabaseIsUpToDate() {
+        var hostHash = fetchHostHash();
+        var localHash = fetchLocalHash();
+
+        // We can't reach the server right now, don't update.
+        if (hostHash == null)
+            return true;
```

Comment:
Maybe throw a custom exception here so that client can tell the user what's going on

Reply:
Actually, I'd like to keep this as a silent failure. This `return true` doesn't abort anything, it just uses the on-disk database if it's there. I think that would enable users to search/list/check for updates on packages when our server is down or our internet isn't quite working.

Reply:
Oh yea right